src/expr/src/relation/func.rs

```diff
@@ -815,6 +815,129 @@ where
     })
 }
 
+// The expected input is in the format of [((OriginalRow, InputValue), OrderByExprs...)]
+fn last_value<'a, I>(
+    datums: I,
+    temp_storage: &'a RowArena,
+    order_by: &[ColumnOrder],
+    window_frame: &WindowFrame,
+) -> Datum<'a>
+where
+    I: IntoIterator<Item = Datum<'a>>,
+{
+    // Sort the datums according to the ORDER BY expressions and return the ((OriginalRow, InputValue), OrderByRow) record
+    // The OrderByRow is kept around because it is required to compute the peer groups in RANGE mode
+    let datums = order_aggregate_datums_with_rank(datums, order_by);
+
+    // Decode the input (OriginalRow, InputValue) into separate datums, while keeping the OrderByRow
+    let datums = datums
+        .into_iter()
+        .map(|(d, order_by_row)| {
+            let mut iter = d.unwrap_list().iter();
+            let original_row = iter.next().unwrap();
+            let input_value = iter.next().unwrap();
+
+            (input_value, original_row, order_by_row)
+        })
+        .collect_vec();
+
+    let length = datums.len();
+    let mut result: Vec<(Datum, Datum)> = Vec::with_capacity(length);
+    for (idx, (current_datum, original_row, order_by_row)) in datums.iter().enumerate() {
+        let last_value = match &window_frame.end_bound {
+            WindowFrameBound::CurrentRow => match &window_frame.units {
+                // Always return the current value when in ROWS mode
+                WindowFrameUnits::Rows => *current_datum,
+                WindowFrameUnits::Range => {
+                    // When in RANGE mode, return the last value of the peer group
+                    // The peer group is the group of rows with the same ORDER BY value
+                    // Note: Range is only supported for the default window frame (RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW),
+                    // which is why it does not appear in the other branches
+                    datums[idx..]
+                        .iter()
+                        .take_while(|(_, _, row)| row == order_by_row)
+                        .last()
+                        .unwrap()
+                        .0
+                }
+                // GROUPS is not supported, and forbidden during planning
+                WindowFrameUnits::Groups => unreachable!(),
+            },
+            WindowFrameBound::UnboundedFollowing => {
+                if let WindowFrameBound::OffsetFollowing(start_offset) = &window_frame.start_bound {
+                    let start_offset = usize::cast_from(*start_offset);
+
+                    // If the frame starts after the last row of the window, return null
+                    if idx + start_offset > length - 1 {
+                        Datum::Null
+                    } else {
+                        datums[length - 1].0
+                    }
+                } else {
+                    datums[length - 1].0
```

Comment:
Is this never called with zero datums?

Reply:
No, as far as I know, because the `for` loop it's in (`datums.iter().enumerate()`) wouldn't execute in that case! I thought about using `datums.last().unwrap()`, but this felt a bit cleaner.

Reply:
Oh, right. Thanks.